File: core/java/src/main/java/hara/lang/protocol/IColl.java
package hara.lang.protocol;

import hara.lang.base.Iter;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;
import java.util.Iterator;

@HaraProtocolBinding(
    namespace = "std.protocol.icoll",
    name = "IColl",
    parents = {"IEquality", "IConj", "IEmpty", "IHash", "IDisplay"})
public interface IColl<E>
    extends Iterable<E>, IEquality, IConj<E>, IEmpty, IHash, IDisplay {

  @HaraMethod(value = "start-string", arity = 1)
  String startString();

  @HaraMethod(value = "end-string", arity = 1)
  String endString();

  @HaraMethod(value = "sep-string", arity = 1)
  default String sepString() {
    return " ";
  }

  @Override
  Iterator<E> iterator();

  @Override
  default String display() {
    return Iter.display(iterator(), startString(), endString(), sepString());
  }
}

File: core/rust/src/lang/protocol/icoll.rs
use super::{IConj, IDisplay, IEmpty, IEquality, IHash};
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.icoll",
    name = "IColl",
    parents = ["IEquality", "IConj", "IEmpty", "IHash", "IDisplay"],
    availability = "portable"
)]
pub trait IColl<E>:
    IntoIterator<Item = E> + IEquality + IConj<E> + IEmpty + IHash + IDisplay
{
    #[hara_method(value = "start-string", arity = 1)]
    fn start_string(&self) -> &'static str;
    #[hara_method(value = "end-string", arity = 1)]
    fn end_string(&self) -> &'static str;

    #[hara_method(value = "sep-string", arity = 1)]
    fn separator(&self) -> &'static str {
        " "
    }
}
